Lockfile generated after `pnpm patch` cannot be used with `--no-frozen-lockfile` on CI for Windows

### pnpm version: 7.4.1

### Code to reproduce the issue:

I [opened a PR](https://github.com/withastro/astro/pull/3747) to replace `patch-package` with `pnpm patch` and `pnpm patch-commit`, but [our CI is failing](https://github.com/withastro/astro/runs/7136424474?check_suite_focus=true) on **Windows** only.

Our lockfile was generated on macOS, so I'm wondering if the paths saved to the lockfile https://github.com/withastro/astro/pull/3747/files#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR3-R10 are the culprit here?

### Expected behavior:

CI environments defaulting to `--no-frozen-lockfile` should still work on Windows.

### Actual behavior:

On Windows only, we're getting this error when running `pnpm install` in CI.

```
ERR_PNPM_FROZEN_LOCKFILE_WITH_OUTDATED_LOCKFILE  Cannot perform a frozen installation because the lockfile needs updates

Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"
```

### Additional information:

 - `node -v` prints: 14, 16
 - Windows, macOS, or Linux?: **Windows** only

--- a/.changeset/soft-seals-shop.md
+++ b/.changeset/soft-seals-shop.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/crypto.base32-hash": patch
+"pnpm": patch
+---
+
+The hash of the patch file should be the same on both Windows and POSIX [#4961](https://github.com/pnpm/pnpm/issues/4961).

--- a/packages/crypto.base32-hash/package.json
+++ b/packages/crypto.base32-hash/package.json
@@ -34,7 +34,8 @@
     "rfc4648": "^1.5.1"
   },
   "devDependencies": {
-    "@pnpm/crypto.base32-hash": "workspace:*"
+    "@pnpm/crypto.base32-hash": "workspace:*",
+    "@pnpm/prepare": "workspace:*"
   },
   "funding": "https://opencollective.com/pnpm",
   "exports": {

--- a/packages/crypto.base32-hash/src/index.ts
+++ b/packages/crypto.base32-hash/src/index.ts
@@ -8,5 +8,5 @@ export function createBase32Hash (str: string): string {
 
 export async function createBase32HashFromFile (file: string): Promise<string> {
   const content = await fs.promises.readFile(file, 'utf8')
-  return createBase32Hash(content.replace('\r\n', '\n'))
+  return createBase32Hash(content.split('\r\n').join('\n'))
 }

--- a/packages/crypto.base32-hash/src/index.ts
+++ b/packages/crypto.base32-hash/src/index.ts
@@ -7,5 +7,6 @@ export function createBase32Hash (str: string): string {
 }
 
 export async function createBase32HashFromFile (file: string): Promise<string> {
-  return createBase32Hash(await fs.promises.readFile(file, 'utf8'))
+  const content = await fs.promises.readFile(file, 'utf8')
+  return createBase32Hash(content.replace('\r\n', '\n'))
 }

--- a/packages/crypto.base32-hash/test/index.ts
+++ b/packages/crypto.base32-hash/test/index.ts
@@ -9,7 +9,7 @@ test('createBase32Hash()', () => {
 
 test('createBase32HashFromFile normalizes line endings before calculating the hash', async () => {
   tempDir()
-  fs.writeFileSync('win-eol.txt', 'a\r\nb')
-  fs.writeFileSync('posix-eol.txt', 'a\nb')
+  fs.writeFileSync('win-eol.txt', 'a\r\nb\r\nc')
+  fs.writeFileSync('posix-eol.txt', 'a\nb\r\nc')
   expect(await createBase32HashFromFile('win-eol.txt')).toEqual(await createBase32HashFromFile('posix-eol.txt'))
 })

--- a/packages/crypto.base32-hash/test/index.ts
+++ b/packages/crypto.base32-hash/test/index.ts
@@ -1,6 +1,15 @@
 /// <reference path="../../../typings/index.d.ts"/>
-import { createBase32Hash } from '@pnpm/crypto.base32-hash'
+import fs from 'fs'
+import { createBase32Hash, createBase32HashFromFile } from '@pnpm/crypto.base32-hash'
+import { tempDir } from '@pnpm/prepare'
 
 test('createBase32Hash()', () => {
   expect(createBase32Hash('AAA')).toEqual('4h5p7m7gcttmf65hikljmi4gw4')
+})
+
+test('createBase32HashFromFile normalizes line endings before calculating the hash', async () => {
+  tempDir()
+  fs.writeFileSync('win-eol.txt', 'a\r\nb')
+  fs.writeFileSync('posix-eol.txt', 'a\nb')
+  expect(await createBase32HashFromFile('win-eol.txt')).toEqual(await createBase32HashFromFile('posix-eol.txt'))
 })

--- a/packages/crypto.base32-hash/tsconfig.json
+++ b/packages/crypto.base32-hash/tsconfig.json
@@ -8,5 +8,9 @@
     "src/**/*.ts",
     "../../typings/**/*.d.ts"
   ],
-  "references": []
+  "references": [
+    {
+      "path": "../../privatePackages/prepare"
+    }
+  ]
 }

